File: source/clibCore/CPCSClass.vb
Imports Contensive.BaseClasses
Imports System.Runtime.InteropServices

Namespace Contensive.Core
    '
    ' comVisible to be activeScript compatible
    '
    <ComVisible(True)> _
    <ComClass(CPCSClass.ClassId, CPCSClass.InterfaceId, CPCSClass.EventsId)> _
    Public Class CPCSClass
        Inherits CPCSBaseClass
        Implements IDisposable
        '
#Region "COM GUIDs"
        Public Const ClassId As String = "63745D9C-795E-4C01-BD6D-4BA35FC4A843"
        Public Const InterfaceId As String = "3F1E7D2E-D697-47A8-A0D3-B625A906BF6A"
        Public Const EventsId As String = "04B8E338-ABB7-44FE-A8DF-2681A36DCA46"
#End Region
        '
        Private cpCore As Contensive.Core.coreClass
        Private CSPointer As Integer
        Private OpeningMemberID As Integer
        Private cp As CPClass
        Protected disposed As Boolean = False
        '
        ' Constructor - Initialize the Main and Csv objects
        '
        public Sub New(ByRef cpParent As CPClass)
            MyBase.New()
            cp = cpParent
            cpCore = cp.core
            CSPointer = -1
            OpeningMemberID = cpCore.user.id
        End Sub
        '
        ' dispose
        '
        Protected Overridable Overloads Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposed Then
                Call appendDebugLog(".dispose, dereference cp, main, csv")
                If disposing Then
                    '
                    ' call .dispose for managed objects
                    '
                    Try
                        If CSPointer <> -1 And True Then
                            Call cpCore.db.cs_Close(CSPointer)
                        End If
                        'If Not (False) Then
                        '    Call cmc.asv.csv_CloseCS(CSPointer)
                        'End If
                    Catch ex As Exception
                    End Try
                    cpCore = Nothing
                    cp = Nothing
                End If
                '
                ' Add code here to release the unmanaged resource.
                '
                'csv = Nothing
            End If
            Me.disposed = True
        End Sub
        '
        ' Insert, called only from Processor41.CSInsert after initializing 
        '
        Public Overrides Function Insert(ByVal ContentName As String) As Boolean
            Dim success As Boolean = False
            '
            Try
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                End If
                CSPointer = cpCore.db.cs_insertRecord(ContentName, OpeningMemberID)
                success = cpCore.db.cs_ok(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.insert")
            End Try
            Return success
        End Function
        '
        '
        '
        Public Overrides Function OpenRecord(ByVal ContentName As String, ByVal recordId As Integer, Optional ByVal SelectFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True) As Boolean
            Dim success As Boolean = False
            '
            Try
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                End If
                CSPointer = cpCore.db.cs_open(ContentName, "id=" & recordId, , ActiveOnly, , , , SelectFieldList, 1, 1)
                success = cpCore.db.cs_ok(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.OpenRecord")
            End Try
            Return success
        End Function
        '
        '
        '
        Public Overrides Function Open(ByVal ContentName As String, Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal SelectFieldList As String = "", Optional ByVal pageSize As Integer = 0, Optional ByVal PageNumber As Integer = 1) As Boolean
            'Public Overrides Function Open(ByVal ContentName As String, Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal SelectFieldList As String = "", Optional ByVal ignore As Integer = 10, Optional ByVal PageNumber As Integer = 1, Optional pageSize As Integer = 0) As Boolean
            Dim success As Boolean = False
            '
            Try
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                End If
                CSPointer = cpCore.db.cs_open(ContentName, SQLCriteria, SortFieldList, ActiveOnly, , , , SelectFieldList, pageSize, PageNumber)
                success = cpCore.db.cs_ok(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.Open")
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function OpenGroupUsers(ByVal GroupList As List(Of String), Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim success As Boolean = False
            '
            Try
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                End If
                CSPointer = cpCore.db.cs_openGroupUsers(GroupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber)
                success = OK()
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.OpenGroupUsers")
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function OpenGroupUsers(ByVal GroupName As String, Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim success As Boolean = False
            '
            Try
                Dim groupList As New List(Of String)
                groupList.Add(GroupName)
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                End If
                CSPointer = cpCore.db.cs_openGroupUsers(groupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber)
                success = OK()
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.OpenGroupUsers")
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function OpenGroupListUsers(ByVal GroupCommaList As String, Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim groupList As New List(Of String)
            '
            groupList.AddRange(GroupCommaList.Split(","))
            OpenGroupListUsers = OpenGroupUsers(groupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function OpenSQL(ByVal sql As String) As Boolean
            Dim success As Boolean = False
            '
            Try
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                End If
                CSPointer = cpCore.db.cs_openCsSql_rev("default", sql)
                success = cpCore.db.cs_ok(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.OpenSQL")
            End Try
            Return success
        End Function
        '
        '
        '
        Public Overrides Function OpenSQL(ByVal sql As String, ByVal DataSourcename As String, Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim success As Boolean = False
            'Dim swap As String
            '
            Try
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                End If
                If ((sql = "") Or (sql.ToLower = "default")) And (DataSourcename <> "") And (DataSourcename.ToLower <> "default") Then
                    '
                    ' support legacy calls were the arguments were was backwards (datasourcename is sql and vise-versa)
                    '
                    CSPointer = cpCore.db.cs_openSql(sql, DataSourcename, PageSize, PageNumber)
                Else
                    CSPointer = cpCore.db.cs_openSql(sql, DataSourcename, PageSize, PageNumber)
                End If
                success = cpCore.db.cs_ok(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.OpenSQL")
            End Try
            Return success
        End Function
        '
        '
        '
        Public Overrides Function OpenSQL2(ByVal sql As String, Optional ByVal DataSourcename As String = "default", Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim success As Boolean = False
            'Dim swap As String
            '
            Try
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                End If
                If ((sql = "") Or (sql.ToLower = "default")) And (DataSourcename <> "") And (DataSourcename.ToLower <> "default") Then
                    '
                    ' support legacy calls were the arguments were was backwards (datasourcename is sql and vise-versa)
                    '
                    CSPointer = cpCore.db.cs_openSql(sql, DataSourcename, PageSize, PageNumber)
                Else
                    CSPointer = cpCore.db.cs_openSql(sql, DataSourcename, PageSize, PageNumber)
                End If
                success = cpCore.db.cs_ok(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.OpenSQL")
            End Try
            Return success
        End Function

        '
        Public Overrides Sub Close()
            Try
                If CSPointer <> -1 Then
                    Call cpCore.db.cs_Close(CSPointer)
                    CSPointer = -1
                End If
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.Close")
            End Try
        End Sub

        Public Overrides Function GetFormInput(ByVal ContentName As String, ByVal FieldName As String, Optional ByVal Height As String = "", Optional ByVal Width As String = "", Optional ByVal HtmlId As String = "") As Object
            If True Then
                Return cpCore.html_GetFormInputCS(CSPointer, ContentName, FieldName, Height, Width, HtmlId)
            Else
                Return ""
            End If
        End Function

        Public Overrides Sub Delete()
            Try
                Call cpCore.db.cs_deleteRecord(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.Delete")
            End Try
        End Sub

        Public Overrides Function FieldOK(ByVal FieldName As String) As Boolean
            Dim result As Boolean = False
            '
            Try
                result = cpCore.db.cs_isFieldSupported(CSPointer, FieldName)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.FieldOK")
            End Try
            Return result
        End Function

        Public Overrides Sub GoFirst()
            Try
                Call cpCore.db.cs_goFirst(CSPointer, False)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.Delete")
            End Try
        End Sub
        '
        '
        '
        Public Overrides Function GetAddLink(Optional ByVal PresetNameValueList As String = "", Optional ByVal AllowPaste As Boolean = False) As String
            Dim result As Object
            '
            Try
                result = cpCore.main_cs_getRecordAddLink(CSPointer, PresetNameValueList, AllowPaste)
                If result Is Nothing Then
                    result = New String("")
                End If
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetAddLink")
                result = New String("")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetBoolean(ByVal FieldName As String) As Boolean
            Dim result As Boolean = False
            '
            Try
                result = cpCore.db.cs_getBoolean(CSPointer, FieldName)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetBoolean")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetDate(ByVal FieldName As String) As Date
            Dim result As Date = #12:00:00 AM#
            '
            Try
                result = cpCore.db.cs_getDate(CSPointer, FieldName)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetDate")
            End Try
            Return result
        End Function

        Public Overrides Function GetEditLink(Optional ByVal AllowCut As Boolean = False) As String
            Dim result As Object
            Try
                result = cpCore.cs_cs_getRecordEditLink(CSPointer, AllowCut)
                If result Is Nothing Then
                    result = New String("")
                End If
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetEditLink")
                result = New String("")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetFilename(ByVal FieldName As String, Optional ByVal OriginalFilename As String = "", Optional ByVal ContentName As String = "") As String
            Dim result As Object
            '
            result = New String("")
            Try
                result = cpCore.db.cs_getFilename(CSPointer, FieldName, OriginalFilename, ContentName)
                If result Is Nothing Then
                    result = New String("")
                End If
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetFilename")
                result = New String("")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetInteger(ByVal FieldName As String) As Integer
            Dim result As Integer = 0
            '
            Try
                result = cpCore.db.cs_getInteger(CSPointer, FieldName)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetInteger")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetNumber(ByVal FieldName As String) As Double
            Dim result As Double = 0
            '
            Try
                result = cpCore.db.cs_getNumber(CSPointer, FieldName)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetNumber")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetRowCount() As Integer
            Dim result As Integer = 0
            '
            Try
                result = cpCore.db.cs_getRowCount(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetRowCount")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetSQL() As String
            Dim result As Object
            '
            result = New String("")
            Try
                result = cpCore.db.cs_getSource(CSPointer)
                If result Is Nothing Then
                    result = New String("")
                End If
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetSQL")
                result = New String("")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetText(ByVal FieldName As String) As String
            Dim result As Object
            '
            result = New String("")
            Try
                result = cpCore.db.cs_get(CSPointer, FieldName)
                If result Is Nothing Then
                    result = New String("")
                End If
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetText")
                result = New String("")
            End Try
            Return result
        End Function
        '
        ' Needs to be implemented (refactor to check the field type. if not fieldTypeHtml, encodeHtml)
        '
        Public Overrides Function GetHtml(ByVal FieldName As String) As String
            Dim result As Object
            result = GetText(FieldName)
            Return result
        End Function
        '
        '
        '
        Public Overrides Function GetTextFile(ByVal FieldName As String) As String
            Dim result As Object
            '
            result = New String("")
            Try
                result = cpCore.db.cs_getTextFile(CSPointer, FieldName)
                If result Is Nothing Then
                    result = New String("")
                End If
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GetTextFile")
                result = New String("")
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Sub GoNext()
            Try
                Call cpCore.db.cs_goNext(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.GoNext")
            End Try
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function NextOK() As Boolean
            Try
                Call cpCore.db.cs_goNext(CSPointer)
                Return cpCore.db.cs_ok(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.NextOK")
            End Try
        End Function
        '
        '
        '
        Public Overrides Function OK() As Boolean
            Dim result As Boolean = False
            '
            Try
                result = cpCore.db.cs_ok(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.OK")
            End Try
            Return result
        End Function
        '
        '
        '
        Public Overrides Sub Save()
            Try
                Call cpCore.db.cs_save2(CSPointer)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.Save")
            End Try
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Sub SetField(ByVal FieldName As String, ByVal FieldValue As Object)
            Try
                Call cpCore.db.cs_set(CSPointer, FieldName, FieldValue)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.SetField")
            End Try
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Sub SetField(ByVal FieldName As String, ByVal FieldValue As String)
            Try
                Call cpCore.db.cs_set(CSPointer, FieldName, FieldValue)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.SetField")
            End Try
        End Sub

        Public Overrides Sub SetFile(ByVal FieldName As String, ByVal Copy As String, ByVal ContentName As String)
            Try
                Call cpCore.db.SetCSTextFile(CSPointer, FieldName, Copy, ContentName)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.SetFile")
            End Try
        End Sub

        Public Overrides Sub SetFormInput(ByVal FieldName As String, Optional ByVal RequestName As String = "")
            Dim success As Boolean = False
            Try
                Call cpCore.cs_setFormInput(CSPointer, FieldName, RequestName)
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in cs.SetFormInput")
            End Try
        End Sub
        Private Sub appendDebugLog(ByVal copy As String)
            ''My.Computer.FileSystem.WriteAllText("c:\clibCpDebug.log", Now & " - cp.cs, " & copy & vbCrLf, True)
            ' 'My.Computer.FileSystem.WriteAllText(System.AppDocmc.main_CurrentDocmc.main_BaseDirectory() & "cpLog.txt", Now & " - " & copy & vbCrLf, True)
        End Sub
        '
        ' testpoint
        '
        Private Sub tp(ByVal msg As String)
            My.Computer.FileSystem.WriteAllText("c:\clibCpTestPoint.log", Now & " - cp.block, " & msg & vbCrLf, True)
        End Sub
#Region " IDisposable Support "
        ' Do not change or add Overridable to these methods.
        ' Put cleanup code in Dispose(ByVal disposing As Boolean).
        Public Overloads Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
        Protected Overrides Sub Finalize()
            Dispose(False)
            MyBase.Finalize()
        End Sub
#End Region
    End Class

End Namespace

File: source/core/Views/cpApi/CPCSClass.cs

using System;
using System.Reflection;
using System.Xml;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using Contensive.Core;
using Contensive.Core.Models.DbModels;
using Contensive.Core.Controllers;
using static Contensive.Core.Controllers.genericController;
using static Contensive.Core.constants;
using Contensive.BaseClasses;
//
namespace Contensive.Core {
    //
    // comVisible to be activeScript compatible
    //
    //[ComVisible(true), Microsoft.VisualBasic.ComClass(CPCSClass.ClassId, CPCSClass.InterfaceId, CPCSClass.EventsId)]
    public class CPCSClass : CPCSBaseClass, IDisposable {
        //
        #region COM GUIDs
        public const string ClassId = "63745D9C-795E-4C01-BD6D-4BA35FC4A843";
        public const string InterfaceId = "3F1E7D2E-D697-47A8-A0D3-B625A906BF6A";
        public const string EventsId = "04B8E338-ABB7-44FE-A8DF-2681A36DCA46";
        #endregion
        //
        private Contensive.Core.coreClass cpCore;
        private int cs;
        private int OpeningMemberID;
        private CPClass cp;
        protected bool disposed = false;
        //
        // Constructor - Initialize the Main and Csv objects
        //
        public CPCSClass(CPClass cpParent) : base() {
            cp = cpParent;
            cpCore = cp.core;
            cs = -1;
            OpeningMemberID = cpCore.doc.sessionContext.user.id;
        }
        //
        // dispose
        //
        protected virtual void Dispose(bool disposing) {
            if (!this.disposed) {
                if (disposing) {
                    try {
                        if (cs != -1 && true) {
                            cpCore.db.csClose(ref cs);
                        }
                    } catch (Exception) {
                    }
                    cpCore = null;
                    cp = null;
                }
            }
            this.disposed = true;
        }
        //
        // Insert, called only from Processor41.CSInsert after initializing 
        //
        public override bool Insert(string ContentName) {
            bool success = false;
            try {
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                }
                cs = cpCore.db.csInsertRecord(ContentName, OpeningMemberID);
                success = cpCore.db.csOk(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //
        //
        public override bool OpenRecord(string ContentName, int recordId, string SelectFieldList = "", bool ActiveOnly = true) {
            bool success = false;
            try {
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                }
                cs = cpCore.db.csOpen(ContentName, "id=" + recordId, "", ActiveOnly, 0, false, false, SelectFieldList, 1, 1);
                success = cpCore.db.csOk(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //
        //
        public override bool Open(string ContentName, string SQLCriteria = "", string SortFieldList = "", bool ActiveOnly = true, string SelectFieldList = "", int pageSize = 10, int PageNumber = 1) {
            bool success = false;
            try {
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                }
                if ((pageSize == 0) || (pageSize == 10)) {
                    // -- (hack) fix for interface issue that has default value 0. later add new method and deprecate
                    // -- pagesize=10, pageNumber=1 -- old code returns all records, new code only returns the first 10 records -- this in effect makes it not compatiblie
                    // -- if I change new cpBase to default pagesize=9999, the change is breaking and old code does not run
                    // -- when I changed new cpbase to pagesize default 0, and compiled code against it, it would not run on c41 because pagesize=0 is passed
                    pageSize = 9999;
                }
                cs = cpCore.db.csOpen(ContentName, SQLCriteria, SortFieldList, ActiveOnly, 0, false, false, SelectFieldList, pageSize, PageNumber);
                success = cpCore.db.csOk(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex); // "Unexpected error in cs.Open") : Throw
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        //
        public override bool OpenGroupUsers(List<string> GroupList, string SQLCriteria = "", string SortFieldList = "", bool ActiveOnly = true, int PageSize = 10, int PageNumber = 1) {
            bool success = false;
            try {
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                }
                cs = cpCore.db.csOpenGroupUsers(GroupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber);
                success = OK();
            } catch (Exception ex) {
                cpCore.handleException(ex); // "Unexpected error in cs.OpenGroupUsers")
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        //
        public override bool OpenGroupUsers(string GroupName, string SQLCriteria = "", string SortFieldList = "", bool ActiveOnly = true, int PageSize = 10, int PageNumber = 1) {
            bool success = false;
            try {
                List<string> groupList = new List<string>();
                groupList.Add(GroupName);
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                }
                cs = cpCore.db.csOpenGroupUsers(groupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber);
                success = OK();
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        //
        [Obsolete()]
        public override bool OpenGroupListUsers(string GroupCommaList, string SQLCriteria = "", string SortFieldList = "", bool ActiveOnly = true, int PageSize = 10, int PageNumber = 1) {
            List<string> groupList = new List<string>();
            groupList.AddRange(GroupCommaList.Split(','));
            return OpenGroupUsers(groupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber);
        }
        //
        //====================================================================================================
        //
        public override bool OpenSQL(string sql) {
            bool success = false;
            try {
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                }
                cs = cpCore.db.csOpenSql_rev("default", sql);
                success = cpCore.db.csOk(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        //
        public override bool OpenSQL(string sql, string DataSourcename, int PageSize = 10, int PageNumber = 1) {
            bool success = false;
            try {
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                }
                if (((string.IsNullOrEmpty(sql)) || (sql.ToLower() == "default")) & (!string.IsNullOrEmpty(DataSourcename)) & (DataSourcename.ToLower() != "default")) {
                    cs = cpCore.db.csOpenSql(sql, DataSourcename, PageSize, PageNumber);
                } else {
                    cs = cpCore.db.csOpenSql(sql, DataSourcename, PageSize, PageNumber);
                }
                success = cpCore.db.csOk(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        //
        public override bool OpenSQL2(string sql, string DataSourcename = "default", int PageSize = 10, int PageNumber = 1) {
            bool success = false;
            try {
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                }
                if (((string.IsNullOrEmpty(sql)) || (sql.ToLower() == "default")) & (!string.IsNullOrEmpty(DataSourcename)) & (DataSourcename.ToLower() != "default")) {
                    cs = cpCore.db.csOpenSql(sql, DataSourcename, PageSize, PageNumber);
                } else {
                    cs = cpCore.db.csOpenSql(sql, DataSourcename, PageSize, PageNumber);
                }
                success = cpCore.db.csOk(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        //
        public override void Close() {
            try {
                if (cs != -1) {
                    cpCore.db.csClose(ref cs);
                    cs = -1;
                }
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override object GetFormInput(string ContentName, string FieldName, string Height = "", string Width = "", string HtmlId = "") {
            return cpCore.html.inputCs(cs, ContentName, FieldName, genericController.encodeInteger(Height), genericController.encodeInteger(Width), HtmlId);
        }
        //
        //====================================================================================================
        //
        public override void Delete() {
            try {
                cpCore.db.csDeleteRecord(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override bool FieldOK(string FieldName) {
            try {
                return cpCore.db.cs_isFieldSupported(cs, FieldName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override void GoFirst() {
            try {
                cpCore.db.cs_goFirst(cs, false);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override string GetAddLink(string PresetNameValueList = "", bool AllowPaste = false) {
            try {
                return cpCore.html.cs_getRecordAddLink(cs, PresetNameValueList, AllowPaste);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override bool GetBoolean(string FieldName) {
            try {
                return cpCore.db.csGetBoolean(cs, FieldName);
            } catch (Exception ex) {
                cpCore.handleException(ex); // "Unexpected error in cs.GetBoolean")
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override DateTime GetDate(string FieldName) {
            try {
                return cpCore.db.csGetDate(cs, FieldName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override string GetEditLink(bool allowCut = false) {
            try {
                return cpCore.db.csGetRecordEditLink(cs, allowCut);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override string GetFilename(string fieldName, string OriginalFilename = "", string ContentName = "") {
            try {
                return cpCore.db.csGetFieldFilename(cs, fieldName, OriginalFilename, ContentName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override int GetInteger(string FieldName) {
            try {
                return cpCore.db.csGetInteger(cs, FieldName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override double GetNumber(string FieldName) {
            try {
                return cpCore.db.csGetNumber(cs, FieldName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override int GetRowCount() {
            try {
                return cpCore.db.csGetRowCount(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override string GetSQL() {
            try {
                return cpCore.db.csGetSource(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override string GetText(string FieldName) {
            try {
                return cpCore.db.csGet(cs, FieldName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override string GetHtml(string FieldName) {
            try {
                return cpCore.db.csGet(cs, FieldName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        [Obsolete("Use getText for copy. getFilename for filename", false)]
        public override string GetTextFile(string FieldName) {
            try {
                return cpCore.db.csGetText(cs, FieldName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override void GoNext() {
            try {
                cpCore.db.csGoNext(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override bool NextOK() {
            try {
                cpCore.db.csGoNext(cs);
                return cpCore.db.csOk(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override bool OK() {
            try {
                return cpCore.db.csOk(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override void Save() {
            try {
                cpCore.db.csSave2(cs);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override void SetField(string FieldName, object FieldValue) {
            try {
                string fieldString = FieldValue as string;
                if (fieldString != null) {
                    cpCore.db.csSet(cs, FieldName, fieldString);
                } else {
                    int? fieldInt = FieldValue as int?;
                    if (fieldInt != null) {
                        cpCore.db.csSet(cs, FieldName, fieldInt.GetValueOrDefault());
                    } else {
                        bool? fieldBool = FieldValue as bool?;
                        if (fieldBool != null) {
                            cpCore.db.csSet(cs, FieldName, fieldBool.GetValueOrDefault());
                        } else {
                            DateTime? fieldDate = FieldValue as DateTime?;
                            if (fieldDate != null) {
                                cpCore.db.csSet(cs, FieldName, fieldDate.GetValueOrDefault());
                            } else {
                                cpCore.db.csSet(cs, FieldName, FieldValue.ToString());
                            }
                        }
                    }
                }
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override void SetField(string FieldName, string FieldValue) {
            try {
                cpCore.db.csSet(cs, FieldName, FieldValue);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override void SetFile(string FieldName, string Copy, string ContentName) {
            try {
                cpCore.db.csSetTextFile(cs, FieldName, Copy, ContentName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        //
        public override void SetFormInput(string FieldName, string RequestName = "") {
            try {
                cpCore.db.cs_setFormInput(cpCore, cs, FieldName, RequestName);
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// Return the value in the field
        /// </summary>
        /// <param name="fieldName"></param>
        /// <returns></returns>
        public override string GetValue(string fieldName) {
            return cpCore.db.cs_getValue(cs, fieldName);
        }

        #region  IDisposable Support 
        // Do not change or add Overridable to these methods.
        // Put cleanup code in Dispose(ByVal disposing As Boolean).
        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~CPCSClass() {
            Dispose(false);
            //todo  NOTE: The base class Finalize method is automatically called from the destructor:
            //base.Finalize();
        }
        #endregion
    }

}

